src/providers/ipa/ipa_deskprofile_rules.c

```diff
@@ -0,0 +1,343 @@
+/*
+    SSSD
+
+    Authors:
+        Fabiano Fidêncio <fidencio@redhat.com>
+
+    Copyright (C) 2017 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "util/util.h"
+#include "providers/ldap/ldap_common.h"
+#include "providers/ldap/sdap_async_private.h"
+#include "providers/ipa/ipa_rules_common.h"
+#include "providers/ipa/ipa_deskprofile_private.h"
+#include "providers/ipa/ipa_deskprofile_rules.h"
+#include "providers/ipa/ipa_deskprofile_rules_util.h"
+
+struct ipa_deskprofile_rule_state {
+    struct tevent_context *ev;
+    struct sdap_handle *sh;
+    struct sdap_options *opts;
+
+    int search_base_iter;
+    struct sdap_search_base **search_bases;
+
+    const char **attrs;
+    char *rules_filter;
+    char *cur_filter;
+
+    size_t rule_count;
+    struct sysdb_attrs **rules;
+};
+
+static errno_t
+ipa_deskprofile_rule_info_next(struct tevent_req *req,
+                               struct ipa_deskprofile_rule_state *state);
+static void
+ipa_deskprofile_rule_info_done(struct tevent_req *subreq);
+
+struct tevent_req *
+ipa_deskprofile_rule_info_send(TALLOC_CTX *mem_ctx,
+                               struct tevent_context *ev,
+                               struct sdap_handle *sh,
+                               struct sdap_options *opts,
+                               struct sdap_search_base **search_bases,
+                               struct sysdb_attrs *ipa_host,
+                               struct sss_domain_info *domain,
+                               const char *username)
+{
+    TALLOC_CTX *tmp_ctx;
+    struct tevent_req *req = NULL;
+    struct ipa_deskprofile_rule_state *state;
+    char *group;
+    char *host_dn_clean;
+    char *group_clean;
+    char *host_group_clean;
+    char *rule_filter;
+    const char *host_dn;
+    const char **memberof_list;
+    char **groups_list;
+    errno_t ret;
+
+    if (ipa_host == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing host\n");
+        return NULL;
+    }
+
+    tmp_ctx = talloc_new(mem_ctx);
```

Comment:
The tevent hierarchy is not correct in this function. In tevent requests, the tmp_ctx is not very often used. Instead, the first thing a `_send` function should do is call `tevent_req_create` to create the state and the request. This is (by convention) the only call that returns NULL from the `send` function. All other calls (because they have a valid request) use the goto done way where the error is marked as failed.

And instead of tmp_ctx, the state is used for temporary data -- because the state hangs off req which hangs off mem_ctx, you always know it's going to be cleaned. In the tevent code, tmp_ctx can be used to avoid memory spikes if some intermediate function needs to allocate a ton of data, but then I would personally allocate tmp_ctx on state anyway.

(Sorry for the brain dump, this should probably be in the coding guidelines or somewhere)